third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp

```diff
@@ -463,574 +299,1153 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+
+    rewriter.replaceOpWithMultiple(whileOp, packedRets);
+
+    return success();
   }
+};
 
-  // Create a new condBranch. We cannot simply extend the operands,
-  // because this would invalidate other operands pointing at the same
-  // cond branch
-  Value tensorPtr = createTensorPointer(fatPtr, curLoc);
-  auto newCondBranch = rewriter.create<cf::CondBranchOp>(
-      curLoc, condBrOp.getCondition(), trueDest, trueOperands, falseDest,
-      falseOperands);
-
-  newCondBranch.setOperand(operandNum, tensorPtr);
-  rewriteOpMap[condBrOp] = newCondBranch;
-  opToDelete.insert(condBrOp);
-  return success();
-}
+/// Rewrite with new operands.
+class ConvertSCFConditionOp
+    : public PointerCanonicalizationPattern<scf::ConditionOp> {
+public:
+  using PointerCanonicalizationPattern::PointerCanonicalizationPattern;
+  LogicalResult
+  matchAndRewrite_(scf::ConditionOp condOp, OneToNOpAdaptor adaptor,
+                   ConversionPatternRewriter &rewriter) const override {
+    SmallVector<Value> newArgs = flattenValues(adaptor.getArgs());
+    // have to mutate here because otherwise scf.while will
+    // get confused about which condition is the "correct" condition (since
+    // there will be two of them before the rewriter DCEs)
+    rewriter.modifyOpInPlace(condOp, [&]() {
+      condOp.getArgsMutable().clear();
+      condOp.getArgsMutable().append(newArgs);
+    });
+    return success();
+  }
+};
+
+/// Rewrite operands for both true dest and false dest.
+class ConvertCFCondBranch
+    : public PointerCanonicalizationPattern<cf::CondBranchOp> {
+public:
+  using PointerCanonicalizationPattern::PointerCanonicalizationPattern;
+  LogicalResult
+  matchAndRewrite_(cf::CondBranchOp branchOp, OneToNOpAdaptor adaptor,
+                   ConversionPatternRewriter &rewriter) const override {
+    ArrayRef<ValueRange> remappedTrueOperands = adaptor.getTrueDestOperands();
+    ArrayRef<ValueRange> remappedFalseOperands = adaptor.getFalseDestOperands();
+    SmallVector<Value> trueOperands = flattenValues(remappedTrueOperands);
+    SmallVector<Value> falseOperands = flattenValues(remappedFalseOperands);
+
+    rewriter.replaceOpWithNewOp<cf::CondBranchOp>(
+        branchOp, branchOp.getCondition(), branchOp.getTrueDest(), trueOperands,
+        branchOp.getFalseDest(), falseOperands);
+
+    // can't put inputNo inside because of limited lifetime (it'll be popped
+    // from stack memory after lambda returns...)
+    auto makeTypeConv = [](unsigned &inputNo,
+                           ArrayRef<ValueRange> remappedOperands) {
+      return [&inputNo, remappedOperands](Type inputType,
+                                          SmallVectorImpl<Type> &types) {
+        SmallVector<Type> remappedInitTypes =
+            llvm::to_vector(remappedOperands[inputNo].getTypes());
+        types.append(remappedInitTypes);
+        inputNo++;
+        return success();
+      };
+    };
+
+    auto propagateCanNarrowToBlock = [this](Block *block,
```

Comment:
Could this be a separate function? I see you are defining this in multiple places, and it looks similar

Reply:
They're all subtly different depending on the op and they're so small that trying to refactor into shared functionality would make them all much more complicated.

Reply:
Went ahead and refactored this for `cf.br`, `cf.cond_br` and `scf.while` but left `scf.for` which is "special".